metal3-scaling-experiments/README.md

```diff
@@ -317,3 +318,63 @@ Caveats:
 
 - The KCP will have some issues since it is not "real", including unknown health for etcd and such.
   This also means that scaling the KCP does not work.
+
+## Debugging with tilt
+
+Tilt-settings:
+
+```yaml
+default_registry: gcr.io/cluster-api-provider
+kind_cluster_name: kind
+enable_providers:
+- kubeadm-bootstrap
+- kubeadm-control-plane
+```
+
+```bash
+# After creating the experiment environment
+export CAPI_KIND_CLUSTER_NAME=kind
+make tilt-up
+```
+
+### Certificate annotation rollout issue
+
+Rollout caused by `MatchesKubeadmBootstrapConfig(machineConfigs, kcp)`
+**The rolling KCPs has kubeadm join configuration *instead* of init configuration.**
+This can be seen by checking the secret containing the kubeadm bootstrap data.
+
+```bash
+kubectl -n test-11 get secret test-11-h7kmr -o jsonpath="{.data.value}" | base64 -d
+# KCPs that rollout has the file /run/kubeadm/kubeadm-join-config.yaml
+# But those that do not rollout instead has /run/kubeadm/kubeadm.yaml
+kubectl -n test-11 get secret test-11-h7kmr -o jsonpath="{.data.value}" | base64 -d | yq '.write_files[] | select(.path == "/run/kubeadm/kubeadm.yaml") | .content' | yq
+kubectl -n test-11 get secret test-11-h7kmr -o jsonpath="{.data.value}" | base64 -d | yq '.write_files[] | select(.path == "/run/kubeadm/kubeadm-join-config.yaml") | .content' | yq
+```
+
+Relevant code: <https://github.com/kubernetes-sigs/cluster-api/blob/87b19f55d71c431b624599d2f3a845ae955a09ee/controlplane/kubeadm/internal/controllers/controller.go#L396-L413>
+
+### Concurrency and Rate Limits
+
+All CAPI controllers has configurable concurrency (default 10), but CAPM3 is lacking this (currently set to 1).
+Relevant code in CAPI: <https://github.com/kubernetes-sigs/cluster-api/blob/d0bd60c23f87a6238a0ae048659ad800e774f81a/controlplane/kubeadm/main.go#L299>
+
+None of the controllers has configurable rate limits.
+The [default in client-go is 10 qps](https://github.com/kubernetes/client-go/blob/02d652e007235a5b46b9972bf136f274983853e6/util/workqueue/default_rate_limiters.go#L39) but the [default in controller-runtime is 20 qps](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.14.5/pkg/client/config/config.go#L96), which is probably where the controllers gets their defaults.
+See also <https://pkg.go.dev/k8s.io/client-go/util/flowcontrol#RateLimiter>
+
+This is likely impacting performance.
+See for example this log from the kubeadm control plane controller:
+
+```text
+I0302 15:50:57.343130       1 request.go:601] Waited for 4.870221494s due to client-side throttling, not priority and fairness, request: GET:https://10.96.0.1:443/api/v1/namespaces/test-697/secrets/test-697-sa
+```
+
+### Custom resources limits and scale targets
+
+See <https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/95-custom-resource-definitions#scale-targets-for-ga>
+
+TODO:
+
```

Comment:
Addition to TODO:

If all the improvements are implemented, someone should redo this study to see how much the speed and resource utilization improved.

Reply:
Done